pkg/arvo/lib/lv.hoon

```diff
@@ -0,0 +1,453 @@
+::
+::::  Vector type in single-precision floating-point @rs
+  ::
+::  Conventions:
+::
+::  m,n,p   always dimensions
+::  i,j,k   always indices (also ii,jj,kk for coder-spec not user-spec)
+::  a,b,c   always lists
+::  u,v,w   always vector/matrix atoms
+::  s,t     always real/floats
+|%
+++  lvs
+  ^|
+  ::~%  %lvs  +>  ~
+  |_  r=$?(%n %u %d %z)   :: round nearest, round up, round down, round to zero
+  ::
+  ::  Manipulators
+  ::
+  ::    Zeroes
+  ++  zeros
+    ::~/  %zeros
+    |=  n=@ud  ^-  @lvs
+    ~_  leaf+"lagoon-fail"
+    `@lvs`(lsh [5 n] 1)   :: pin at head for leading zeros
+  ::
+  ::    Fill value
+  ++  fill
+    |=  [n=@ud s=@rs]  ^-  @lvs
+    `@lvs`(mix (zeros n) (fil 5 n s))
+  ::
+  ::    Ones
+  ++  ones
+    |=  n=@ud  ^-  @lvs
+    ~_  leaf+"lagoon-fail"
+    (fill n .1)
+  ::
+  ::    Length of vector
+  ++  length
+    |=  u=@lvs  ^-  @ud
+    ~_  leaf+"lagoon-fail"
+    =/  ell  (met 5 u)
+    ?:  (gth ell 1)  (dec ell)  0
+  ::
+  ::    Produce a vector from `(list @u)` (of natural numbers)
+  ++  make-nat
+    |=  a=(list @u)  ^-  @lvs
+    (make (turn a sun:rs))
+  ::
+  ::    APL-style index list
+  ++  iota
+    |=  n=@u  ^-  @lvs
+    (make-nat (gulf 1 n))
+  ++  make
+    |=  [a=(list @rs)]  ^-  @lvs
+    ~_  leaf+"lagoon-fail"
+    `@lvs`(mix (rep [5 1] a) (zeros (lent a)))
+  ++  unmake
+    |=  [u=@lvs]  ^-  (list @rs)
+    ~_  leaf+"lagoon-fail"
+    (flop `(list @rs)`+:(flop (rip 5 u)))
+  ++  append
+    |=  [u=@lvs s=@rs]  ^-  @lvs
+    (make (snoc (unmake u) s))
+    ::  XX could be done faster with a mix/lsh
+  ::
+  ::  Yield the substring [lhs:rhs] inclusive
+  ++  subvector
+    |=  [u=@lvs lhs=@ud rhs=@ud]
+    (mix (zeros +((sub rhs lhs))) (cut 5 [(dec lhs) +((sub rhs lhs))] u))
+  ::
+  ::  |x|
+  ++  abs
+    |=  [s=@rs]
+    ?:  (gth:rs s .0)  s  (sub:rs .0 s)
+  ::
+  ::  |x-y| <= tol
+  ++  isclose
+    |=  [s=@rs t=@rs tol=@rs]
+    (lth:rs (abs (sub:rs s t)) tol)
+  ++  near0
+    |=  s=@rs
+    (isclose s .0 .1e-6)
+  ++  all-close
+    |=  [u=@lvs v=@lvs tol=@rs]
+      =/  n  `@ud`(length u)
+      =/  count  1
+      =/  off  0
+      |-
+        ?:  (gth off 0)  %.n
+        ?:  (gth count n)  %.y
```

Comment:
`%.y` and `%.n` are constant atoms -- their type is their value. `?` is `?(%.y %.n)`. The type of `&` or `|` is `?` (you can also use `.y` and `.n`, but idk if I've ever seen those). This is an awkward corner of the language, and also the problem you're hitting in your tests.